adding Ising Model example

AroosaIjaz opened the pull request:
Adding the Ising model examples to the example folder on Pennylane. 

There is a detailed jupyter file and three corresponding python files to compare the same example across different interfaces. 

This example uses and compares different optimizers whilst also showing limitations of optimizers on non-convex cost functions.

quantshah commented:
Hi, thanks for the PR. Looks very interesting and maybe I can help review it. I have some suggestions on the overall style. Check this refactoring of your Q-isingmodel-PL.py. I think something like this will render better independently, e.g., https://nipy.org/dipy/examples_built/reconst_ivim.html#example-reconst-ivim

```
"""
===========
Ising model
===========
Example of qubit optimization for three qubit Ising model in qml default qubit
ecosystem. We will be comparing `pennylane.optimize.GradientDescentOptimizer`
and `tensorflow.train.GradientDescentOptimizer` for this quantum system.
"""

import pennylane as qml
from pennylane import numpy as np
from pennylane.optimize import GradientDescentOptimizer

"""
First, we make a device backend with the backend plugin as `default.qubit`.
We will be solving the problem with a fixed coupling between the three qubits
and optimizing the phase operator to get minimum energy configuration of the
spin system.

Let us set the first spin to be always up (+1 eigenstate of Pauli-Z operator).
Then we can optimize the rotation angles for the other two spins so that the
energy in minimized for the given couplings. We can use the general
Rot(phi, theta, omega, wires) single qubit operation.
"""

dev1 = qml.device("default.qubit", wires=3)


@qml.qnode(dev1)
def circuit1(p1, p2):
    qml.Rot(p1[0], p1[1], p1[2], wires=1)
    qml.Rot(p2[0], p2[1], p2[2], wires=2)
    return qml.expval.PauliZ(0), qml.expval.PauliZ(1), qml.expval.PauliZ(2)


"""
Assume that the coupling matrix is `J = [1, -1]`. `circuit1` function returns
a numpy array of Pauli-Z expectation values, i.e, +1 for spin up and -1 for
spin down. The cost function gives the Ising energy for a given state,
"""


def cost(var):
    J = np.array([1, -1])
    spins = circuit1(var[0], var[1])
    energy = -sum(J_ij * spins[i] * spins[i + 1] for i, J_ij in enumerate(J))
    return energy


"""
Let us test it for the spin configuration [1, -1, -1]. The total energy for
this ising model should be
 .. math::
    H = -1(J_1 s_1 s_2 + J_2 s_2 s_3) = 2 
"""

# t1=np.array([0, np.pi, 0])
# t2=np.array([0,np.pi ,0])
t1 = np.pi * (np.random.ranf(3))
t2 = np.pi * (np.random.ranf(3))
var_init = np.array([t1, t2])
cost_init = cost(var_init)
print(var_init)
print(cost_init)

"""
Now we can optimize the parameters using the pennylane optimizer, 
`pennylane.optimize.GradientDescentOptimizer`
"""

gd = GradientDescentOptimizer(0.4)

var = var_init
var_gd = [var]
cost_gd = [cost_init]

for it in range(100):
    var = gd.step(cost, var)
    if (it + 1) % 5 == 0:
        var_gd.append(var)
        cost_gd.append(cost(var))
        print("Energy after step {:5d}: {: .7f} | Angles: {}".format(
                it + 1, cost(var), var))

"""
As expected, the minimum energy is -2 for the spin configuration [1, 1, -1]
which corresponds to (phi, theta, omega = (0,0,0)) for the second spin and (0, pi, 0)
for the third spin.
"""
cost_final = cost(var)
print(cost_final)

```

josh146 commented:
Hi @AroosaIjaz, thanks for these great set of examples! It looks fantastic. Just a couple of small notes and suggestions before doing the full code review.

1. To match the filenames of our other examples, could you rename the files to `Q5_isingmodel-NumPy.py`, `Q5_isingmodel-PyTorch.py`, `Q5_isingmodel-TF.py`, etc.

2. Run the black code formatter on your Python scripts. This can be done by `pip install black`, and then running `black filename.py -l 100`

3. Delete any commented lines of code, as they are not being used

I'm in favour of the changes suggested by @quantshah, however it is best to continue to use Python comments for descriptions, rather than multi-line strings (as multi-line strings are executable Python statements).

AroosaIjaz commented:
Thank you, @josh146  and @quantshah . 

I cleaned up the files and made the changes you suggested. Let me know If everything looks okay now. 

Thanks!

josh146 commented:
Thanks @AroosaIjaz! Looks great, I'll post a code-review either today or tomorrow (also, remember to `git rm` the old files from this branch, with the previous filenames)

AroosaIjaz commented:
Thanks for reminding me! I forgot to remove them!

josh146 commented on examples/Q5_isingmodel-NumPy.py:
```diff
@@ -0,0 +1,88 @@
+# Qubit Optimization for 3 qubit Ising model using  Pennylane default qubit ecosystem.
```
```suggestion
# Qubit Optimization for 3 qubit Ising model using the Pennylane 'default.qubit' plugin.
```

josh146 commented on examples/Q5_isingmodel-NumPy.py:
```diff
@@ -0,0 +1,88 @@
+
+# Ising models are used in Quantum Annealing to solve Quadratic Unconstrained Binary
+
+# Optimization problems with non-convex  cost functions. This example demonstrates how
```
```suggestion
# Optimization problems with non-convex cost functions. This example demonstrates how
```

josh146 commented on examples/Q5_isingmodel-NumPy.py:
```diff
@@ -0,0 +1,88 @@
+
+# This example  uses and compares pennylane gradient descent optimizer with the optimizers
+
+# of PyTorch and and TensorFlow for this quantum system: see Q5_isingmodel.ipynb
```
looks great! Might be a bit easier to read by removing the double line spacing here

josh146 commented on examples/Q5_isingmodel-NumPy.py:
```diff
@@ -0,0 +1,88 @@
+from pennylane import numpy as np
+from pennylane.optimize import GradientDescentOptimizer
+
+# making a device with pennylane dafault qubit command
```
```suggestion
# making a PennyLane device using the 'default.qubit' plugin
```

josh146 commented on examples/Q5_isingmodel-NumPy.py:
```diff
@@ -0,0 +1,88 @@
+
+@qml.qnode(dev1)
+def circuit1(p1, p2):
+    # 	lets say first spin is always up (+1 eigenstate of pauli-z operator)
```
```suggestion
    # 	Assuming the first spin is always up (+1 eigenstate of the Pauli-Z operator)
```

josh146 commented on examples/Q5_isingmodel-NumPy.py:
```diff
@@ -0,0 +1,88 @@
+def cost(var):
+    # 	let coupling matrix be J=[1, -1]
+    J = np.array([1, -1])
+    #       circuit1 function returns a numpy array of pauliz exp values
```
```suggestion
    # circuit1 function returns a numpy array of pauliz exp values
```

josh146 commented on examples/Q5_isingmodel-NumPy.py:
```diff
@@ -0,0 +1,88 @@
+test1 = np.array([0, np.pi, 0])
+test2 = np.array([0, np.pi, 0])
+cost_check = cost([test1, test2])
+print(cost_check)
```
It might be nice to include more information in these print statement for the user in the terminal, something like
```python
print("Energy for [1, -1, -1] spin configuration:", cost_check)
```
(This can also be done for all following print statements)

josh146 commented on examples/Q5_isingmodel-NumPy.py:
```diff
@@ -0,0 +1,88 @@
+    if (it + 1) % 5 == 0:
+        var_gd.append(var)
+        cost_gd.append(cost(var))
+        print("Energy after step {:5d}: {: .7f} | Angles: {}".format(it + 1, cost(var), var))
```
Nice!

josh146 commented on examples/Q5_isingmodel-NumPy.py:
```diff
@@ -0,0 +1,88 @@
+print(var_init)
+print(cost_init)
+
+# Now we optimize using pennylane numpy gradient descent optimizer
```
```suggestion
# Now we optimize using the PennyLane NumPy radient descent optimizer
```

AroosaIjaz commented:
Thanks a lot @josh146 for the careful review! I have incorporated all these changes

quantshah commented:
This looks ready to merge. Can you remove the two junkfile.txt? @AroosaIjaz

AroosaIjaz commented:
@quantshah thanks I will have a look. We made the junkfiles on the first day in some branch. Seems like a year ago already :) I will look it up  :)

AroosaIjaz commented on examples/Q5_isingmodel.ipynb:
```diff
@@ -0,0 +1,606 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In some runs, you will see the optimization getting stuck in a local minima of energy (usually zero energy). Do not forget to restart the kernel before rerunning the file! "
```
yeah. I can add more detail here in the notebook. 
In every run they might see different optimization paths as the intialization is random for all the optimizers. If you rerun the file (after restarting the kernel-so it does not effect the new run), you might see different results.

AroosaIjaz replied:
Have a look at the edited file and let me know if you think its better. 

Thanks!

AroosaIjaz commented:
This is good for a final review!

co9olguy commented:
Hi @AroosaIjaz, I was going to do second review on this, but I notice that there are still a large number of suggested changes from the previous round of review that have not been accepted/resolved. Can you take a look and resolve those please?

AroosaIjaz commented:
Oh I think I missed them somehow. Will check again. My apologies

AroosaIjaz commented on examples/Q5_isingmodel-NumPy.py:
```diff
@@ -0,0 +1,84 @@
+dev1 = qml.device("default.qubit", wires=3)
+
+# we will be solving the problem with fixed coupling between the three
+# qubits and optimizing the phase operator to get minimum energy
```
yep there is one possible minimum energy for the fixed coupling matrix we are using.

co9olguy replied:
ok, but the sentence still needs an article. Based on your comment above, I suppose it is "the minimum energy"

AroosaIjaz replied:
yeah. I agree

AroosaIjaz commented on examples/Q5_isingmodel-PyTorch.py:
```diff
@@ -0,0 +1,92 @@
+print("Randomly initialized angles: ",var_init)
+print("Corresponding cost before initialization:  ",cost_init)
+
+# Now we use PyTorch optimizer to reduce this cost
```
Not really. I think I saw Adam optimizer being used in other examples involving PyTorch as well on the website. so kept it consistent.

AroosaIjaz commented:
Now, it should be ready for a final review :)

co9olguy commented on examples/Q5_isingmodel-NumPy.py:
```diff
@@ -0,0 +1,84 @@
+dev1 = qml.device("default.qubit", wires=3)
+
+# we will be solving the problem with fixed coupling between the three
+# qubits and optimizing the phase operator to get minimum energy
```
```suggestion
# qubits and optimizing the phase operator to get the minimum energy
```

co9olguy commented on examples/Q5_isingmodel.ipynb:
```diff
@@ -0,0 +1,606 @@
+    "    qml.Rot(p2[0],p2[1],p2[2], wires =2)\n",
+    "    return qml.expval.PauliZ(0), qml.expval.PauliZ(1), qml.expval.PauliZ(2) \n",
+    "\n",
+    "# the expectation value of pauliZ is plus 1 for spin up and -1 for spin down."
```
Any reason this was marked resolved with no change when the same suggested change elsewhere was accepted?

AroosaIjaz replied:
My apologies. I think I did not see it properly as the comment fonts are harder to read...I fixed it now. I will be more careful with these when programming from now on.

co9olguy commented on examples/Q5_isingmodel-PyTorch.py:
```diff
@@ -0,0 +1,92 @@
+p1_final, p2_final = opt.param_groups[0]["params"]
+
+# the minimum energy is -2  for the spin configuration [1,1,-1] which corresponds to
+# (phi,theta,omega = (0,0,0) for spin2 and (0,pi,0) for spin3
```
```suggestion
# (phi,theta,omega) = (0,0,0) for spin2 and (0,pi,0) for spin3
```

co9olguy commented on examples/Q5_isingmodel-NumPy.py:
```diff
@@ -0,0 +1,84 @@
+
+
+# the minimum energy is -2  for the spin configuration [1,1,-1] which corresponds to
+# (phi,theta,omega = (0,0,0) for spin2 and (0,pi,0) for spin3
```
```suggestion
# (phi,theta,omega) = (0,0,0) for spin2 and (0,pi,0) for spin3
```

co9olguy commented on examples/Q5_isingmodel-TF.py:
```diff
@@ -0,0 +1,90 @@
+        print("Energy after step {:5d}: {: .7f} | Angles: {}".format(i + 1, cost([t1, t2]), [t1, t2]),"\n")
+
+# the minimum energy is -2  for the spin configuration [1,1,-1] which corresponds to
+# (phi,theta,omega = (0,0,0) for spin2 and (0,pi,0) for spin3
```
```suggestion
# (phi,theta,omega) = (0,0,0) for spin2 and (0,pi,0) for spin3
```

co9olguy commented on examples/Q5_isingmodel.ipynb:
```diff
@@ -0,0 +1,666 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Qubit Optimization for 3-qubit ising model. \n",
```
```suggestion
    "# Qubit Optimization for 3-qubit Ising model. \n",
```

co9olguy commented on examples/Q5_isingmodel.ipynb:
```diff
@@ -0,0 +1,666 @@
+   "source": [
+    "Let's test these functions for [1,-1,-1] spin configuration. \n",
+    "\n",
+    "Total energy for this ising model should be\n",
```
```suggestion
    "Total energy for this Ising model should be\n",
```

co9olguy commented on examples/Q5_isingmodel.ipynb:
```diff
@@ -0,0 +1,666 @@
+   "source": [
+    "# Qubit Optimization for 3-qubit ising model. \n",
+    "\n",
+    "We will be comparing the PennyLane gradient descent optimizer with the optimizers in Tensorflow and PyTorch for this quantum system. The Ising model is an example of a system whose optimization landscape is non-convex and hence using gradient descent may not be the best strategy since the optimization can get stuck in local minimas. Quantum annealing (for example on D-wave system) is one alt ...
```
```suggestion
    "We will be comparing the PennyLane gradient descent optimizer with the optimizers in Tensorflow and PyTorch for this quantum system. The Ising model is an example of a system whose optimization landscape is non-convex and hence using gradient descent may not be the best strategy since the optimization can get stuck in local minimas. Quantum annealing (for example on a D-wave system) is one alternative that may have more success in finding the global ground state. \n",
```

co9olguy commented on examples/Q5_isingmodel.ipynb:
```diff
@@ -0,0 +1,666 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let's test these functions for [1,-1,-1] spin configuration. \n",
```
```suggestion
    "Let's test these functions for the [1,-1,-1] spin configuration. \n",
```

co9olguy commented on examples/Q5_isingmodel.ipynb:
```diff
@@ -0,0 +1,666 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The minimun energy should be -2 for [1,1,-1] spin configuration which corresponds to the angles of [0,0,0] and [0,pi,0] for second and third spin respectively. "
```
```suggestion
    "The minimun energy should be -2 for the [1,1,-1] spin configuration which corresponds to the angles of [0,0,0] and [0,pi,0] for second and third spin respectively. "
```

AroosaIjaz commented:
> (i) we should find and fix random seeds which we know lead to both local and global minima. This would be more illustrative (and productive) than asking users to rerun the example code many times and hope they see one of the other outcomes
 
Yeah! I agree. Really nice suggestion. :)

AroosaIjaz commented:
@co9olguy I have added seeds and changed the Adam optimizer to the SGD (stochastic-GD) optimizer for PyTorch. I dont think PyTorch has a batch-GD optimizer. I am not sure but I think PennyLane and TensorFlow are using batch-GD optimization.

AroosaIjaz commented:
@co9olguy I have added the the three python files (with modifications) to the new tutorial documentation. We can perhaps just add the jupyter file of this example with this PR. As far as I understand, the jupyter files stay on GitHub only.

co9olguy commented:
Thanks @AroosaIjaz, we're going to fix a new release of PennyLane with recent code changes first, then go through and review/merge all the waiting documentation PRs shortly afterwards
